src/common/provider.ts

```diff
@@ -1,10 +1,43 @@
-import { JsonRpcProvider } from '@ethersproject/providers';
+import { providers, utils } from 'ethers';
 import { ConfigService } from '@nestjs/config';
 import { Injectable } from '@nestjs/common';
+import { Logger } from '../modules/logger/logger.service';
+
+const { poll } = utils;
 
 @Injectable()
-export class Provider extends JsonRpcProvider {
-  constructor(configService: ConfigService) {
+export class Provider extends providers.JsonRpcProvider {
+  private attempts: number;
+
+  constructor(configService: ConfigService, private logger: Logger) {
     super(configService.get('ENS_URL'));
+    this.attempts = configService.get('CHAIN_REQUEST_ATTEMPTS');
+  }
+
+  /**
+   * Retries failed chain requests https://github.com/ethers-io/ethers.js/issues/427#issuecomment-465329448
+   * @override
+   */
+  public perform(method: string, params: any) {
+    let attempts = 0;
+    return poll(() => {
+      attempts++;
+      return super.perform(method, params).then(
+        (result) => {
+          return result;
+        },
+        (error) => {
+          if (attempts >= this.attempts) {
+            this.logger.warn(
+              `Failed to ${method} with params ${params} after ${this.attempts}`
+            );
+            return error;
```

Comment:
This is going to resolve the promise returned by the `perform` method. Is this expected or should rejected promise be returned here?

Reply:
Ah, right. Error should be thrown in order to reject